File: lintcode/Combination_Sum.cpp
class Solution {
public:
    /**
     * @param candidates: A list of integers
     * @param target:An integer
     * @return: A list of lists of integers
     */
    vector<vector<int> > combinationSum(vector<int> &candidates, int target) {
        if(candidates.empty()) return {};
        
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> result;
        vector<int> sofar;
        
        combinationHelper(result, sofar, candidates, 0, target);
        return result;
    }
private:
    void combinationHelper(vector<vector<int>> &result, vector<int> &sofar, vector<int> &remain, int k, int target){
        int sum = accumulate(sofar.begin(), sofar.end(), 0);
        if(sum == target) {
            result.push_back(sofar);
            return;
        }
        if(k == remain.size()) return;
        
        int x = remain[k];
        int t = 0;
        while(sum + t*x <= target){
            if(t > 0) sofar.push_back(x);
            combinationHelper(result, sofar, remain, k+1, target);
            t++;
        }
        
        while(t > 1){
            sofar.pop_back();
            t--;
        }
    }
};

File: lintcode/java/CombinationSum.java
public class Solution {
    /**
     * @param candidates: A list of integers
     * @param target:An integer
     * @return: A list of lists of integers
     */
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        if(candidates == null || candidates.length == 0) return Collections.emptyList();
        
        Arrays.sort(candidates);
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> sofar = new ArrayList<>();
        
        combinataionSumHelper(result, sofar, candidates, 0, target);
        return result;
    }
    
    private void combinataionSumHelper(List<List<Integer>> result, List<Integer> sofar, int[] candidates, int ind, int target){
        int sum = sum(sofar);
        if(sum == target){
            result.add(new ArrayList<>(sofar));
            return;
        }
        if(sum > target || ind ==candidates.length) return;
        
        int num = candidates[ind];
        int cnt = 0;
        for(;sum <= target; sum+=num){
            if(cnt > 0) sofar.add(num);
            combinataionSumHelper(result, sofar, candidates, ind+1, target);
            cnt++;
        }

        while(cnt > 1){
            sofar.remove(sofar.size()-1);
            cnt--;
        }
    }
    
    private int sum(List<Integer> nums){
        int sum = 0;
        for(int x: nums) sum+=x;
        return sum;
    }
}
